providers/azure_test.go

```diff
@@ -122,82 +123,96 @@ func testAzureBackend(payload string) *httptest.Server {
-	defer b.Close()
-
-	bURL, _ := url.Parse(b.URL)
-	p := testAzureProvider(bURL.Host)
-
-	session := CreateAuthorizedSession()
-	email, err := p.GetEmailAddress(context.Background(), session)
-	assert.Equal(t, nil, err)
-	assert.Equal(t, "", email)
-}
-
-func TestAzureProviderGetEmailAddressIncorrectOtherMails(t *testing.T) {
-	b := testAzureBackend(`{ "mail": null, "otherMails": "", "userPrincipalName": null }`)
-	defer b.Close()
-
-	bURL, _ := url.Parse(b.URL)
-	p := testAzureProvider(bURL.Host)
+func TestAzureProviderEnrichSessionState(t *testing.T) {
+	testCases := []struct {
+		Description             string
+		IDToken                 string
+		PayloadFromAzureBackend string
+		ExpectedEmail           string
+		ExpectedError           error
+	}{
+		{
+			Description:             "EnrichSessionState should return email using mail property from Azure backend",
+			PayloadFromAzureBackend: `{ "mail": "user@windows.net" }`,
+			ExpectedEmail:           "user@windows.net",
+			ExpectedError:           nil,
+		},
+		{
+			Description:             "EnrichSessionState should return email using otherMails property returned from Azure backend",
+			PayloadFromAzureBackend: `{ "mail": null, "otherMails": ["user@windows.net", "altuser@windows.net"] }`,
+			ExpectedEmail:           "user@windows.net",
+			ExpectedError:           nil,
+		},
+		{
+			Description:             "EnrichSessionState should return email using userPrincipalName from Azure backend",
+			PayloadFromAzureBackend: `{ "mail": null, "otherMails": [], "userPrincipalName": "user@windows.net" }`,
+			ExpectedEmail:           "user@windows.net",
+			ExpectedError:           nil,
+		},
+		{
+			Description:             "EnrichSessionState should return error when Azure backend doesn't return email information",
+			PayloadFromAzureBackend: `{ "mail": null, "otherMails": [], "userPrincipalName": null }`,
+			ExpectedEmail:           "",
+			ExpectedError:           errors.New("type assertion to string failed"),
+		},
+		{
+			Description:             "EnrichSessionState should return empty email when userPrincipalName from Azure backend is empty",
+			PayloadFromAzureBackend: `{ "mail": null, "otherMails": [], "userPrincipalName": "" }`,
+			ExpectedEmail:           "",
+			ExpectedError:           nil,
+		},
+		{
+			Description:             "EnrichSessionState should return error when otherMails from Azure backend is not a valid type",
+			PayloadFromAzureBackend: `{ "mail": null, "otherMails": "", "userPrincipalName": null }`,
+			ExpectedEmail:           "",
+			ExpectedError:           errors.New("type assertion to string failed"),
+		},
+		{
+			Description: "EnrichSessionState should return email from id_token",
+			// { "sub": "1234567890", "email": "foo@bar.com", "iat": 1516239022 }
+			IDToken:       "eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJzdWIiOiIxMjM0NTY3ODkwIiwiZW1haWwiOiJmb29AYmFyLmNvbSIsImlhdCI6MTUxNjIzOTAyMn0.XRuL4Y2VPSToNB8vMvmlB-X3BwahUJzUXNx6vmzODjk",
```

Comment:
Before I forget before we pause this work for a bit:

For these test cases can we get mock IDTokens that look like what Azure would give back (headers + payloads, sig can be garbage).

That will make testing compatibility with Azure OIDC changes a lot easier for other contributors

Reply:
I believe we will need to properly "generate and sign" these token claims once oidc is implemented in the provider.

Reply:
You don't have to expose a real token & signature of yours.  You can rip up the insides and replace the payload claims object with mock values and re-encode.

We'd do a setup like this in unit tests where we don't actually verify the signature, we just use a mock JWKS: https://github.com/oauth2-proxy/oauth2-proxy/blob/3ff0c23a9e418875cdd7460b0a6a5b40dea71770/providers/oidc_test.go#L110